ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/ImmDB.hs

```diff
@@ -293,6 +296,24 @@ appendBlock db@ImmDB{..} b = withDB db $ \imm -> case isEBB b of
   Streaming
 -------------------------------------------------------------------------------}
 
+-- | Wrapper around 'ImmDB.streamBinaryBlobs' that 'register's a call to
+-- 'ImmDB.iteratorClose' as a clean-up action in the 'ResourceRegistry'.
+--
+-- When the returned iterator is closed, it will also be 'unregister'ed from
+-- the 'ResourceRegistry'.
+registeredStream :: (MonadMask m, MonadSTM m, HasHeader blk)
+                 => ImmDB m blk
+                 -> ResourceRegistry m
+                 -> Maybe (SlotNo, HeaderHash blk)
+                 -> Maybe (SlotNo, HeaderHash blk)
+                 -> m (ImmDB.Iterator (HeaderHash blk) m Lazy.ByteString)
+registeredStream db registry start end = mask $ \unmask -> do
+    it  <- unmask $ withDB db $ \imm -> ImmDB.streamBinaryBlobs imm start end
+    key <- register registry (iteratorClose db it)
+    return it
```

Comment:
Let's add a comment here. Adding this to the close action of the iterator is not crucial for correctness, but it avoids the resource registry growing too much.

Reply:
Also, let's introduce the equivalent of [allocate](http://hackage.haskell.org/package/resourcet-1.2.2/docs/Control-Monad-Trans-Resource.html#v:allocate) so that this `mask` dance happens once and for all in the resource registry.